### PR TITLE
feat: advanced filters

### PR DESCRIPTION
**Pull request by michael-0acf4**

### CODE REVIEW

File: substantial/backends/redis.py [hunk 1]
```diff
@@ -11,70 +9,133 @@
 class RedisBackend(Backend):
     def __init__(self, host: str, port: int, **kwargs):
         self.redis = redis.Redis(host=host, port=port, decode_responses=True, **kwargs)
+        self.separator = ":/"
```
Comment (zifeo) [inline substantial/backends/redis.py:12]: any specific reason for 2 chars?

File: substantial/backends/redis.py [hunk 2]
```diff
@@ -11,70 +9,133 @@
...
     async def read_events(self, run_id: str) -> Union[Records, None]:
-        key = "_".join(["runs", run_id, "events"])
-        val = self.redis.get(key)
+        event_key = self._key("runs", run_id, "events")
+        val: str | None = self.redis.register_script("""
+            local event_key = KEYS[1]
+            if redis.call("EXISTS", event_key) == 1 then
+                return redis.call("GET", event_key)
+            else
+                return nil
```
Comment (zifeo) [inline substantial/backends/redis.py:42]: doesn't get already return nil when the key doesn't exist?

File: substantial/backends/redis.py [hunk 3]
```diff
@@ -11,70 +9,133 @@
...
-            ret.append(log)
-        return ret
+        log_key = self._key("runs", run_id, "logs")
+        logs = self.redis.register_script("""
+            local log_key = KEYS[1]
+            local run_id = ARGV[1]
+
+            local sched_keys = redis.call("LRANGE", log_key, 0, -1)
+            local logs = {}
+            for _, sched_key in ipairs(sched_keys) do
```
Comment (zifeo) [inline substantial/backends/redis.py:60]: when you use Lua you make concurrent request wait, I wonder wether a lrange + mget would not be a better fit here. Similarly for the follow ups once, I would rely on Lua only if there is no other command required

File: substantial/filters.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+class Ok:
+    value: any
+
+
+@dataclass
+class Err:
+    value: any
+
+
+Result = Union[Ok, Err, None]
```
Comment (zifeo) [inline substantial/filters.py:21]: 💯

File: substantial/backends/backend.py [hunk 5]
```diff
@@ -5,20 +5,26 @@
 
 
 class Backend:
-    # run related
+    # metadata related
+    async def read_all_metadata(self, run_id: str) -> List[Metadata]:
```
Comment (zifeo) [inline substantial/backends/backend.py:9]: can we document the key-value structure and role of each one? I am not 100% clear about the new link ones and their actual needs after reading the PR.

**Review by zifeo — COMMENTED** [review]
Nice work!

File: substantial/backends/backend.py (hunk 5 above)
Comment (michael-0acf4) [reply substantial/backends/backend.py:9]: Very sorry about that, but I guess I will leave a quick summary for the time being:
The only addition on this PR is basically adding a link from workflow name to runs directly on the underlying backend, others are refactor work. It is basically just metadata so we can answer "what are all runs related to a given workflow name" (name => List run). A similar pattern is also used for child workflows on metatype (name => List name).

File: substantial/backends/redis.py (hunk 2 above)
Comment (michael-0acf4) [reply substantial/backends/redis.py:42]: Good point, I don't quite remember but I think I had conversion error or something like that from lua world to rust, also here I pretty much wanted to sync both versions as I did some improvements on the [metatype version](https://github.com/metatypedev/metatype/blob/main/src/substantial/src/backends/redis.rs).

Could be an outdated issue though.

File: substantial/backends/redis.py (hunk 3 above)
Comment (michael-0acf4) [reply substantial/backends/redis.py:60]: > when you use Lua you make concurrent request wait
This happens on a specific run but not on all of them so I was not really worried on the performance. The other complex lua scripts follow similar rationals.

> I would rely on Lua only if there is no other command required
I think it still matters if multiple wites are involved or if in between two value retrieval there is no delete from outside.

> lrange + mget
That could actually work! Also this is all read ops without potential deletes in between queries so we can get rid of Lua in this situation I guess.

File: substantial/backends/redis.py (hunk 1 above)
Comment (michael-0acf4) [reply substantial/backends/redis.py:12]: I had issues with `:` only, also a single character as a separator has higher risk of unintentionally splitting meaningful parts of a key (e.g., `2024-01-01 00:00:00` when using `sep=':'`).
I wanted to keep  `:` as it is supported by many Redis client browsers like [this one](https://github.com/joeferner/redis-commander) for grouping the keys so I just added `/`.

**Review by zifeo — APPROVED** [review]

**Pull request merged**